File: packages/sdk/src/integrations/langgraph.ts
import { BaseCheckpointSaver, Checkpoint, CheckpointMetadata, CheckpointTuple } from "@langchain/langgraph";
import { RunnableConfig } from "@langchain/core/runnables";
import { LetsPing } from "../index";

type StoredCheckpoint = {
    checkpoint: Checkpoint;
    metadata: CheckpointMetadata;
};

export class LetsPingCheckpointer extends BaseCheckpointSaver {
    private checkpoints: Record<string, StoredCheckpoint> = {};

    constructor(public client: LetsPing) {
        super();
    }

    private getTransport(): (<T = any>(method: string, path: string, body?: any) => Promise<T>) | null {
        const clientAny = this.client as any;
        if (typeof clientAny.request === "function") {
            return clientAny.request.bind(this.client);
        }
        return null;
    }

    private async saveRemote(
        threadId: string,
        checkpointId: string,
        checkpoint: Checkpoint,
        metadata: CheckpointMetadata
    ): Promise<void> {
        const transport = this.getTransport();
        if (!transport) {
            console.warn("[LetsPingCheckpointer] Missing underlying transport; falling back to in-memory only.");
            return;
        }
        try {
            await transport("POST", "/langgraph/checkpoints", {
                thread_id: threadId,
                checkpoint_id: checkpointId,
                checkpoint,
                metadata,
            });
        } catch (e) {
            console.warn("[LetsPingCheckpointer] Failed to persist checkpoint remotely; falling back to in-memory only.", e);
        }
    }

    private async loadRemote(
        threadId: string,
        checkpointId?: string
    ): Promise<StoredCheckpoint | null> {
        const transport = this.getTransport();
        if (!transport) {
            console.warn("[LetsPingCheckpointer] Missing underlying transport; using in-memory checkpoints only.");
            return null;
        }
        const search = checkpointId
            ? `?thread_id=${encodeURIComponent(threadId)}&checkpoint_id=${encodeURIComponent(checkpointId)}`
            : `?thread_id=${encodeURIComponent(threadId)}&latest=1`;
        try {
            const res = await transport<any>("GET", `/langgraph/checkpoints${search}`);
            if (res && res.checkpoint && res.metadata) {
                return { checkpoint: res.checkpoint as Checkpoint, metadata: res.metadata as CheckpointMetadata };
            }
        } catch (e) {
            // If not found or backend unavailable, fall back to local cache only.
            console.warn("[LetsPingCheckpointer] Failed to load remote checkpoint", e);
        }
        return null;
    }

    private async deleteRemote(threadId: string): Promise<void> {
        const transport = this.getTransport();
        if (!transport) return;
        const search = `?thread_id=${encodeURIComponent(threadId)}`;
        try {
            await transport("DELETE", `/langgraph/checkpoints${search}`);
        } catch (e) {
            console.warn("[LetsPingCheckpointer] Failed to delete remote checkpoints", e);
        }
    }

    async put(
        config: RunnableConfig,
        checkpoint: Checkpoint,
        metadata: CheckpointMetadata,
        newVersions?: Record<string, string | number>
    ): Promise<RunnableConfig> {
        const threadId = config.configurable?.thread_id;
        const checkpointId = checkpoint.id;

        if (!threadId || !checkpointId) {
            return config;
        }

        this.checkpoints[`${threadId}:${checkpointId}`] = { checkpoint, metadata };
        await this.saveRemote(threadId, checkpointId, checkpoint, metadata);

        return {
            configurable: {
                thread_id: threadId,
                checkpoint_id: checkpointId,
            },
        };
    }

    // METHODS REQUIRED BY LANGGRAPH V0.1+
    async putWrites(config: RunnableConfig, writes: any, taskId: string): Promise<void> {
        // No-op for V1: LetsPing focuses on primary state parking, not granular sub-task writes.
    }

    async deleteThread(threadId: string): Promise<void> {
        for (const key of Object.keys(this.checkpoints)) {
            if (key.startsWith(`${threadId}:`)) {
                delete this.checkpoints[key];
            }
        }
        await this.deleteRemote(threadId);
    }

    async getTuple(config: RunnableConfig): Promise<CheckpointTuple | undefined> {
        const threadId = config.configurable?.thread_id;
        const checkpointId = config.configurable?.checkpoint_id;
        if (!threadId) return undefined;

        // Prefer remote truth, fall back to local cache.
        const remote = await this.loadRemote(threadId, checkpointId);
        if (remote) {
            return { config, checkpoint: remote.checkpoint, metadata: remote.metadata };
        }

        if (checkpointId) {
            const match = this.checkpoints[`${threadId}:${checkpointId}`];
            if (match) {
                return { config, checkpoint: match.checkpoint, metadata: match.metadata };
            }
        }

        let latest: CheckpointTuple | undefined;
        for (const [key, val] of Object.entries(this.checkpoints)) {
            if (key.startsWith(`${threadId}:`)) {
                latest = { config, checkpoint: val.checkpoint, metadata: val.metadata };
            }
        }
        return latest;
    }

    async *list(config: RunnableConfig, options?: any): AsyncGenerator<CheckpointTuple> {
        const threadId = config.configurable?.thread_id;
        if (!threadId) return;

        const remoteLatest = await this.loadRemote(threadId);
        if (remoteLatest) {
            yield { config, checkpoint: remoteLatest.checkpoint, metadata: remoteLatest.metadata };
        }

        for (const [key, val] of Object.entries(this.checkpoints)) {
            if (key.startsWith(`${threadId}:`)) {
                yield { config, checkpoint: val.checkpoint, metadata: val.metadata };
            }
        }
    }
}

File: packages/python/letsping/integrations/langgraph.py
# packages/python/letsping/integrations/langgraph.py
from typing import Optional, Any, Dict, Iterator, Tuple

from langchain_core.runnables import RunnableConfig
from langgraph.checkpoint.base import BaseCheckpointSaver, Checkpoint, CheckpointMetadata, CheckpointTuple
from .. import LetsPing


class LetsPingCheckpointer(BaseCheckpointSaver):
    """
    A LangGraph Checkpointer that persists checkpoints remotely via the LetsPing control plane.

    Checkpoints are encrypted and stored alongside your existing Cryo-Sleep state in Supabase
    Storage, scoped per project + thread.
    """

    def __init__(self, client: LetsPing):
        super().__init__()
        self.client = client
        self._checkpoints: Dict[str, Tuple[Checkpoint, CheckpointMetadata]] = {}

    def _key(self, thread_id: str, checkpoint_id: str) -> str:
        return f"{thread_id}:{checkpoint_id}"

    def _save_remote(
        self,
        thread_id: str,
        checkpoint_id: str,
        checkpoint: Checkpoint,
        metadata: CheckpointMetadata,
    ) -> None:
        body = {
            "thread_id": thread_id,
            "checkpoint_id": checkpoint_id,
            "checkpoint": checkpoint,
            "metadata": metadata,
        }
        try:
            resp = self.client._client.post("langgraph/checkpoints", json=body)
            self.client._handle_response(resp)
        except Exception as e:
            # Remote persistence is best-effort; LangGraph will still function with in-memory checkpoints only.
            from .. import logger  # reuse SDK logger
            logger.warning(f"LetsPingCheckpointer: failed to persist remote checkpoint, falling back to memory-only: {e}")

    def _load_remote(
        self,
        thread_id: str,
        checkpoint_id: Optional[str] = None,
    ) -> Optional[Tuple[Checkpoint, CheckpointMetadata]]:
        params: Dict[str, Any] = {"thread_id": thread_id}
        if checkpoint_id:
            params["checkpoint_id"] = checkpoint_id
        else:
            params["latest"] = "1"

        try:
            resp = self.client._client.get("langgraph/checkpoints", params=params)
            if resp.status_code == 404:
                return None
            data = self.client._handle_response(resp)
            cp = data.get("checkpoint")
            meta = data.get("metadata")
            if cp is None or meta is None:
                return None
            return cp, meta
        except Exception as e:
            from .. import logger
            logger.warning(f"LetsPingCheckpointer: failed to load remote checkpoint, using memory-only: {e}")
            return None

    def _delete_remote(self, thread_id: str) -> None:
        resp = self.client._client.delete("langgraph/checkpoints", params={"thread_id": thread_id})
        # Ignore specific errors; this is best-effort cleanup.
        if resp.status_code not in (200, 204, 404):
            self.client._handle_response(resp)

    def put(
        self,
        config: RunnableConfig,
        checkpoint: Checkpoint,
        metadata: CheckpointMetadata,
        new_versions: dict[str, str | float | int],
    ) -> RunnableConfig:
        thread_id = config["configurable"].get("thread_id")
        checkpoint_id = checkpoint.get("id")

        if thread_id is None or checkpoint_id is None:
            return config

        self._checkpoints[self._key(thread_id, checkpoint_id)] = (checkpoint, metadata)
        self._save_remote(thread_id, checkpoint_id, checkpoint, metadata)

        return {
            "configurable": {
                "thread_id": thread_id,
                "checkpoint_id": checkpoint_id,
            }
        }

    def get_tuple(self, config: RunnableConfig) -> Optional[CheckpointTuple]:
        thread_id = config["configurable"].get("thread_id")
        checkpoint_id = config["configurable"].get("checkpoint_id")

        if thread_id is None:
            return None

        # Prefer the remote source of truth, fall back to local cache.
        remote = self._load_remote(thread_id, checkpoint_id)
        if remote is not None:
            cp, meta = remote
            return CheckpointTuple(config, cp, meta)

        if checkpoint_id:
            key = self._key(thread_id, checkpoint_id)
            if key in self._checkpoints:
                cp, meta = self._checkpoints[key]
                return CheckpointTuple(config, cp, meta)

        latest: Optional[CheckpointTuple] = None
        prefix = f"{thread_id}:"
        for key, (cp, meta) in self._checkpoints.items():
            if key.startswith(prefix):
                latest = CheckpointTuple(config, cp, meta)
        return latest

    def list(
        self,
        config: Optional[RunnableConfig],
        *,
        filter: Optional[Dict[str, Any]] = None,
        before: Optional[RunnableConfig] = None,
        limit: Optional[int] = None,
    ) -> Iterator[CheckpointTuple]:
        if not config:
            return iter(())

        thread_id = config["configurable"].get("thread_id")
        if thread_id is None:
            return iter(())

        results: list[CheckpointTuple] = []

        # Best-effort: include the latest remote checkpoint first.
        remote = self._load_remote(thread_id)
        if remote is not None:
            cp, meta = remote
            results.append(CheckpointTuple(config, cp, meta))

        prefix = f"{thread_id}:"
        for key, (cp, meta) in self._checkpoints.items():
            if key.startswith(prefix):
                results.append(CheckpointTuple(config, cp, meta))

        if limit is not None:
            results = results[:limit]
        return iter(results)